### PR TITLE
filter for Open File panel

### LINKED ISSUE #32: The Open File panel should filter out non-PDF files

### PR DESCRIPTION
**Pull request by Ra-V2**
Fixes #32

### CODE REVIEW

**Comment by owenca** [comment]
A couple of issues:
- Dangling symlinks should be filtered.
- Symlinks to non-PDF files should be filtered.

File: application/PDFFilter.cpp [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+    		nameLink.ToUpper();
+    		int32 l = nameLink.FindLast('.');
+    		if (l != B_ERROR) {
+    			if (nameLink.FindFirst(".PDF",1) != B_ERROR) return true;
+    			else if (nameLink.FindFirst(".DJVU",1) != B_ERROR) return true;
+    			else if (nameLink.FindFirst(".XPS",1) != B_ERROR) return true;
+    		}
+		}
+    return false;
+	}
```
Comment (owenca) [inline application/PDFFilter.cpp]:
```
	for (int i = 0; valid_filetypes[i] != nullptr; i++)
		// if mime type matches, return true

	return false;
```

File: application/PDFFilter.cpp [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+ * Authors:
+ *		Ciprian Nedisan (cipri)
+ *
+ */
+ 
+#include <cstring>
+#include <String.h>
+#include <SymLink.h>
+
+#include "PDFFilter.h"
```
Comment (owenca) [inline application/PDFFilter.cpp]:
```
#include "PDFFilter.h"

#include <string.h>

File: application/PDFFilter.cpp [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+ * Authors:
+ *		Ciprian Nedisan (cipri)
+ *
+ */
+
+#include "PDFFilter.h"
+
+#include <MimeType.h>
+#include <string.h>
+#include <SymLink.h>
```
Comment (owenca) [inline application/PDFFilter.cpp]: [Haiku coding guidelines](https://www.haiku-os.org/development/coding-guidelines/):
```
#include <string.h>

#include <MimeType.h>
```

File: application/PDFFilter.cpp [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+static const char* valid_filetypes[] = {
+	"application/pdf",
+	"application/djvu",
+	"application/xps",
+	NULL
+};
+
+bool
+PDFFilter::Filter(const entry_ref *ref, BNode *node, struct stat_beos *st,
+	const char *filetype)
```
Comment (owenca) [inline application/PDFFilter.cpp]: Haiku coding guidelines, as pointed out [before](https://github.com/HaikuArchives/DocumentViewer/pull/35#discussion_r361353658).

File: application/PDFFilter.cpp [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+{
+	BEntry entry(ref, true);
```
Comment (owenca) [inline application/PDFFilter.cpp]:
```suggestion
	BEntry entry(ref, true);
	if (entry.InitCheck() != B_OK)
		return false;
	if (entry.IsDirectory())
		return true;
```

File: application/PDFFilter.cpp [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+	if(strcmp(filetype, "application/x-vnd.Be-directory") == 0) return true;
+	else if(strcmp(filetype, "application/x-vnd.Be-volume") == 0) return true;
+	else if(strcmp(filetype, "application/x-vnd.Be-symlink") == 0) {
+		BSymLink symLink(ref);
+		char buffer[256];
+		symLink.ReadLink(buffer, sizeof(buffer));
+		BMimeType::GuessMimeType(buffer, &mimeType);
+	}
+	else
+		BMimeType::GuessMimeType(ref, &mimeType);
```
Comment (owenca) [inline application/PDFFilter.cpp]:
```
	entry_ref linkedRef;
	if (entry.GetRef(&linkedRef) != B_OK)
		return false;

	BMimeType mimeType;
	if (BMimeType::GuessMimeType(&linkedRef, &mimeType) != B_OK)
		return false;
```

File: application/PDFFilter.cpp [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+};
+
+bool
+PDFFilter::Filter(const entry_ref *ref, BNode *node, struct stat_beos *st,
+	const char *filetype)
+{
+	BEntry entry(ref, true);
+	BMimeType mimeType;
+
+	if(!entry.Exists()) return false;
```
Comment (owenca) [inline application/PDFFilter.cpp]: Remove.

File: application/PDFFilter.cpp [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+		BSymLink symLink(ref);
+		char buffer[256];
+		symLink.ReadLink(buffer, sizeof(buffer));
+		BMimeType::GuessMimeType(buffer, &mimeType);
+	}
+	else
+		BMimeType::GuessMimeType(ref, &mimeType);
+
+	for (int i = 0; valid_filetypes[i] != nullptr; i++)
+    	if(strcmp(mimeType.Type(), valid_filetypes[i]) == 0) return true;
```
Comment (owenca) [inline application/PDFFilter.cpp]:
```suggestion
		if (strcmp(mimeType.Type(), valid_filetypes[i]) == 0)
			return true;
```

**Comment by owenca** [comment]
Your [repo](https://github.com/Ra-V2/DocumentViewer) is still 1 commit behind. You can fix it thus:
```
git pull --rebase upstream/master
# Fix the whitespace warning
git commit -a --amend
git push -f
```

**Comment by humdingerb** [comment]
Thanks!

**Pull request merged**